### PR TITLE
Add `Host`/`HostInterface` to C++

### LINKED ISSUE #331: Add `Host`/`HostInterface` to C++
## What

The mirror of `Manager`/`ManagerInterface` for the host representation.

## ACs

- `Host` class in the `managerAPI` C++ namespace.
- `HostInterface` class in the `hostAPI` C++ namespace (incl. `HostInterfacePtr`).
- Support for:
  - `identifier`
  - `displayName`
  - `info`
- Issue for `C` implementation, with any design notes.
- Python bindings as the base for existing Python `Host`/`HostInterface` classes
- Existing `test_host.py` tests used for coverage
   - Additions for the handling of edge cases where invalid values are supplied to constructors/etc.

## Out of scope

 - `_interface`
 - loggin/audit

### PR DESCRIPTION
**Pull request by feltech**
Closes #331

### CODE REVIEW

File: src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * This may be used by a Manager's @ref
+   * openassetio.managerAPI.ManagerInterface "ManagerInterface" to
+   * adjust its behavior accordingly. The identifier should be
+   * unique for any application, but common to all versions.
+   *
+   * The identifier should use only alpha-numeric characters and '.',
+   * '_' or '-'. We suggest using the "reverse DNS" style, for
+   * example:
+   *
+   *    "com.foundry.katana"
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp:59]: [fix]
```suggestion
   *    "org.openassetio.host.test"
```

File: src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+   *
+   * @return host identifier.
+   */
+  [[nodiscard]] virtual openassetio::Str identifier() const = 0;
+
+  /**
+   * Returns a human readable name to be used to reference this
+   * specific host in user-facing presentations.
+   *
+   *     "Katana"
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp:69]: [fix]
```suggestion
   *     "OpenAssetIO Test Host"
```

File: src/openassetio-core/include/openassetio/managerAPI/Host.hpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  /**
+   * Returns an identifier to uniquely identify the Host.
+   *
+   * The identifier will be different for each tool or application,
+   * but common to all versions of any one. The identifier will use
+   * only alpha-numeric characters and '.', '_' or '-', commonly in
+   * the form of a 'reverse-DNS' style string, for example:
+   *
+   *     "uk.co.foundry.katana"
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/managerAPI/Host.hpp:47]: [fix]
```suggestion
   *     "org.openassetio.host.test"
```

File: src/openassetio-core/include/openassetio/managerAPI/Host.hpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  [[nodiscard]] Str identifier() const;
+
+  /**
+   * Returns a human readable name to be used to reference this
+   * specific host in user-facing messaging.
+   * For example:
+   *
+   *     "Katana"
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/managerAPI/Host.hpp:56]: [fix]
```suggestion
   *     "OpenAssetIO Test Host"
```

**Review by foundrytom — CHANGES_REQUESTED** [review]
Just a few minor tweaks.

**Comment by feltech** [comment]
I took the liberty of squashing and rewording the commits. 

The final two commits are new and could do with a review. Specifically 64fdfde4 and 14ff361e.

File: tests/python/openassetio/hostAPI/test_session.py [hunk 5]
```diff
@@ -153,6 +130,23 @@ def test_when_useManager_was_called_with_valid_identifier_then_initializes_expec
         mock_manager_factory.instantiate.assert_called_once_with(an_id)
         mock_manager_interface.mock.initialize.assert_called_once()
         mock_manager_interface.mock.setSettings.assert_not_called()
+        # Assert provided host session contains the expected host. We
+        # assume correct behaviour of Host methods, i.e. that they
+        # delegate to the HostInterface as appropriate.
+        # I.e. we set HostInterface.identifier() to some known value,
+        # then:
```
Comment (foundrytom) [inline tests/python/openassetio/hostAPI/test_session.py:137]:
```suggestion
        # then check that the host returned by the HostSession, captured
        # by our StubManager during initialization (above) returns this
        # same value:
```

**Comment by feltech** [comment]
> [fix] Release notes use past tense.

Hopefully addressed in 7bb8e5c8 and d18c5bb7 (?)

File: src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  /**
+   * Returns other information that may be useful about this Host.
+   * This can contain arbitrary key/value pairs. Managers never rely
+   * directly on any particular keys being set here, but the
+   * information may be useful for diagnostic or debugging purposes.
+   * For example:
+   *
+   * { 'version' : '1.1v3' }
+   *
+   *     @return Dict[str, pod]
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp:84]: [fix] I think this needs removing now?

File: src/openassetio-python/managerAPI/HostBinding.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+#include <openassetio/managerAPI/Host.hpp>
+
+#include "../_openassetio.hpp"
+
+void registerHost(const py::module& mod) {
+  using openassetio::hostAPI::HostInterfacePtr;
+  using openassetio::managerAPI::Host;
+
+  // Host wrapper is cheap and has no independent shared state, so
+  // no need for a `shared_ptr` holder.
```
Comment (foundrytom) [inline src/openassetio-python/managerAPI/HostBinding.cpp:14]: [question] Out of date comment?

File: tests/python/openassetio/managerAPI/test_host.py [hunk 8]
```diff
@@ -41,20 +41,22 @@ def a_context():
...
             Host(object())
 
         assert str(err.value).startswith("__init__(): incompatible constructor arguments")
 
 
-class Test_Host_interface:
-    def test(self, mock_host_interface):
-        # pylint: disable=protected-access
-        a_host = Host(mock_host_interface)
-        assert a_host._interface() is mock_host_interface
```
Comment (foundrytom) [inline tests/python/openassetio/managerAPI/test_host.py:56]: [consider] Do we still want some kind of unit test for this? We could set some test-custom return value for the identifier a method of our mock, call that method and check its returned by `host.identifier()`?

**Review by foundrytom — CHANGES_REQUESTED** [review]
Thanks @feltech, looking good! Just a few small things.

File: tests/python/openassetio/managerAPI/test_host.py (hunk 8 above)
Comment (feltech) [reply tests/python/openassetio/managerAPI/test_host.py:56]: There is already `Test_Host_identifier` (and other methods) that check the call is forwarded to the provided `HostInterface`.

File: src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp (hunk 6 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/hostAPI/HostInterface.hpp]: Reworded in f8a4a46c

File: tests/python/openassetio/managerAPI/test_host.py (hunk 8 above)
Comment (foundrytom) [reply tests/python/openassetio/managerAPI/test_host.py]: Ahh perfect.

**Review by foundrytom — APPROVED** [review]
Thanks @feltech, feel free to squash and merge at your leisure.
